### PR TITLE
virtual-workspaces: wire into kcp server and remove topology leaking to cli plugin

### PR DESCRIPTION
**Pull request by sttts**
Adds
```
--run-virtual-workspaces=true/false # Run the virtual workspace apiservers in-process (defaults to true!)
--virtual-workspace-address=<url> # Address of a stand-alone virtual workspace apiservers
--virtual-workspaces-workspaces-base-path # The prefix of the workspaces API server root path.
```
to `kcp start`.

Note: `virtual-workspaces-workspaces-base-path` is not a typo 🤷‍♂️ 

In case of `--run-virtual-workspaces=false` it requires `--virtual-workspace-address` to set a redirect (301) URL. Obviously in a production environment with uniform URLs, this can be taken over by a front-proxy (cc @kylape @csams).

Among the in-process run mode, this also closes https://github.com/kcp-dev/kcp/issues/638.

TODOs:
- [ ] adapt demo to use in-process mode
- [ ] adapt services deployment @csams @kylape

### CODE REVIEW

File: test/e2e/virtual/workspaces/virtual_workspace_test.go [hunk 1]
```diff
@@ -76,38 +82,52 @@ var testData = testDataType{
 }
 
 func TestWorkspacesVirtualWorkspaces(t *testing.T) {
```
Comment (marun) [inline test/e2e/virtual/workspaces/virtual_workspace_test.go:84]: Do we need Parallel here too?

Comment (sttts) [reply test/e2e/virtual/workspaces/virtual_workspace_test.go:84]: doesn't harm, no?

But of course follow-up: use one fixture. 6 of them is little insane.

Comment (sttts) [reply test/e2e/virtual/workspaces/virtual_workspace_test.go:84]: fixed

File: test/e2e/virtual/workspaces/virtual_workspace_test.go [hunk 2]
```diff
@@ -298,14 +320,26 @@ func TestWorkspacesVirtualWorkspaces(t *testing.T) {
 			usersKCPArgs, err := framework.Users(users).ArgsForKCP(t)
 			require.NoError(t, err)
 
+			// create port early. We have to hope it is still free when we are ready to start the virtual workspace apiserver.
+			portStr, err := framework.GetFreePort(t)
```
Comment (marun) [inline test/e2e/virtual/workspaces/virtual_workspace_test.go:324]: (No action required) The port allocation performed by GetFreePort is problematic to me. There is no guarantee that the port we allocate will remain free - a non-test entity not bound by our file-based locking scheme could easily end up using our chosen port before our process using it starts. Ideally anything relying on this strategy for port allocation would at least retry.

Even better - a process would accept a port of 0 and enable discovery of the port that it gets allocated (e.g. by writing it to a known location on disk, kubeconfig, etc). I always wondered why kube didn't support this directly.

File: test/e2e/virtual/workspaces/virtual_workspace_test.go [hunk 3]
```diff
@@ -321,75 +355,102 @@ func TestWorkspacesVirtualWorkspaces(t *testing.T) {
...
-					// write kubeconfig to disk, next to kcp kubeconfig
-					cfgPath := filepath.Join(filepath.Dir(kcpServer.KubeconfigPath()), "virtualworkspace.kubeconfig")
-					err = clientcmd.WriteToFile(virtualWorkspaceKubeConfig, cfgPath)
+					},
+					AuthInfos: map[string]*clientcmdapi.AuthInfo{
+						"virtualworkspace": kcpAdminConfig.AuthInfos["admin"],
+					},
+					CurrentContext: "shard",
+				}
+				kubeconfgiPath := filepath.Join(filepath.Dir(server.KubeconfigPath()), "virtualworkspace.kubeconfig")
```
Comment (marun) [inline test/e2e/virtual/workspaces/virtual_workspace_test.go:377]: nit: s/kubeconfgi/kubeconfig/

File: pkg/virtual/framework/architecture.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+  1. common authentication
+  2. common authorization
+  3. common OpenAPI endpoint
+  4. all APIs under the same /apis/<group>.
+  5. common discovery endpoints
+
+In other words, an apiserver delegation chain serves one workspace only.
+
+For virtual workspace apiserver in contrast, we need to server multiple workspaces,
```
Comment (marun) [inline pkg/virtual/framework/architecture.md:16]: nit: s/server/serve/

File: pkg/virtual/framework/architecture.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  5. independent discovery endpoints <some-prefix>/apis.
+
+On the other hand, we want to keep using k8s.io/apiserver. To make this work with
+the upper requirements, the handler chain will 
+
+  a) inject the virtual workspace into the context
+  b) rewrite the request path to remove the virtual workspace prefix.
+  
+The delegated apiserver will wrap the s.Handler's director (which usually decides
+which requests to pass to goretful) and check for the virtual workspace name in the
```
Comment (marun) [inline pkg/virtual/framework/architecture.md:32]: goretful?

File: pkg/virtual/framework/architecture.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+context. If it matches, the request is forwarded to the original directory, and
```
Comment (marun) [inline pkg/virtual/framework/architecture.md:33]: directory?

File: test/e2e/virtual/workspaces/virtual_workspace_test.go (hunk 2 above)
Comment (sttts) [reply test/e2e/virtual/workspaces/virtual_workspace_test.go:324]: the apiserver library allows that in-process. You can instantiate a listener manually with kernel port zero, which means to choose and bind one. Then you can do whatever you want and then start up the server with that listener. This obviously does not work out-of-process.

Comment (marun) [reply test/e2e/virtual/workspaces/virtual_workspace_test.go:324]: I'm saying it *should* work out-of-process, provided a way to communicate the port e.g. via fs.

File: pkg/virtual/framework/architecture.md (hunk 5 above)
Comment (sttts) [reply pkg/virtual/framework/architecture.md:32]: gorestful.

File: pkg/virtual/framework/architecture.md (hunk 6 above)
Comment (sttts) [reply pkg/virtual/framework/architecture.md:33]: director

**Comment by marun** [comment]
Test changes LGTM

File: cmd/virtual-workspaces/command/cmd.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+	"github.com/kcp-dev/kcp/cmd/virtual-workspaces/options"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	kcpinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	"github.com/kcp-dev/kcp/pkg/virtual/framework"
+	virtualrootapiserver "github.com/kcp-dev/kcp/pkg/virtual/framework/rootapiserver"
+)
+
+type SubCommandOptions interface {
+	NewVirtualWorkspaces() ([]virtualrootapiserver.InformerStart, []framework.VirtualWorkspace, error)
```
Comment (davidfestal) [inline cmd/virtual-workspaces/command/cmd.go:49]: This file seems to mix things related to all virtual workspaces (like `SubCommandOptions`) with things related to the `workspaces` virtual workspace (the one managing personal workspaces).
I'd rather have command definitions for each virtual workspace in a sub-folder

**Review by davidfestal — CHANGES_REQUESTED** [review]

File: cmd/virtual-workspaces/command/cmd.go (hunk 7 above)
Comment (sttts) [reply cmd/virtual-workspaces/command/cmd.go:49]: It does intentionally until we have another command.

File: cmd/virtual-workspaces/options/options.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	genericapiserveroptions "k8s.io/apiserver/pkg/server/options"
+
+	workspacesoptions "github.com/kcp-dev/kcp/pkg/virtual/workspaces/options"
+)
+
+type Options struct {
```
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/virtual-workspaces/options/options.go:39]: took another look. I don't think that statement is true. These options should be generic for all virtual workspaces. Only what is different among them goes e.g. into workspaceoptions.Workspaces.

[the remark this answers is not in the document]
Comment (davidfestal) [reply cmd/virtual-workspaces/options/options.go:39]: Yeah, that's right that, now, things are in a much better shape to allow sharing a single kubeconfig among the various virtual workspaces

**Review by davidfestal — APPROVED** [review]
Great work ! Thanks !

**Pull request merged**